### DISCUSSION TITLE
Einsy USB to serial firmware issues vs RPI Zero 2 w

### DISCUSSION THREAD

**Discussion by Malvix** [dz0ny/klipper-prusa-mk3s#8] [Q&A] [answered]
I noticed on the generic Einsy klipper config provided on the klipper repo that apparently "The Einsy boards sold by Prusa have defective firmware on the usb-to-serial chip that make the boards unusable with Klipper". 

Given that in your instructions you used a RPI Zero 2 w via GPIO (which iirc is a serial connection), I'm assuming this configuration doesn't suffer from the same firmware issues? Or did you have to flash a fixed USB to serial firmware on top of the steps outlined? Thanks!

**Comment by dz0ny** [comment, marked as answer]
No, no extra flashing is not needed you are using one of the free serial ports not usb to serial port. On that note I also had no issues going USB way before fully committing to conversion, I have mk3s from 2020 so I beleive they don't have issue anymore.

**Reply by confuzled-dev** [reply]
I apologize for resurrecting this post but I am planning to install klipper and was worried about this same issue. Is there a way to test is I have the issue that needs flashing.

My mk3s is from January 2021.

I am a bit nervous but looking forward to things, just a bit concerned with how old most of the guides are so if I can follow your steps and not worry about the serial flashing, that would be a huge help

**Reply by dz0ny** [reply]
If your HW has an issue with the USB as described in the guides, klipper won't even work as it uses a much faster baud rate. 

Basically, you would see something like "MCU did not respond in time to the command" from the Klipper when trying to connect via USB. In that case, you can just use the exposed UART as it's described in the Prusa Connect article. 

TLDR; No flashing is required, Klipper works just fine via USB or via UART.

**Reply by confuzled-dev** [reply]
I am still a bit unsure so please forgive my ignorance.

I thought the pi was going to be contacted to the USB port on the prusa control box.

There was a einsy bug but it no longer applies to this setup?

**Reply by dz0ny** [reply]
You have two options:

a) permanent via pins on board https://help.prusa3d.com/article/prusaprint-rpi-zero-and-octoprint_2180
b) usb

> There was a einsy bug but it no longer applies to this setup?

There was yes, I have never observed it. From reports only those who used Octoprint observed it.

**Reply by confuzled-dev** [reply]
dzony, I have successfully installed mainsailOS to my host device and flashed Klipper to my Prusa. I noticed some parts that seemed unclear which caused a bit of confusion on my end. If I submitted a write-up on possible changes/suggestions to make the process a bit more approachable for a new user, would you be interested?

**Comment by confuzled-dev** [comment]
Ah, so if I only plan to use MainsailOS (for example) it shouldn't be an
issue

Edit: and to be clear, I plan to use a raspberry pi 3 not a zero

**Comment by JesterIsDead** [comment]
dougshell, I did have the USB issue but was able to fix it and had klipper running via USB. Via GPI0 I have no connection to MCU. Anyluck on your end? I'm using a banana pi m2 zero, but I think it should be the same as they are the same GPI0 pins.

**Comment by confuzled-dev** [comment]
I apologize but I only partial follow.

I haven't installed klipper yet as I am working on a project. I hope to
flash the einsy fix tonight or tomorrow and attempt klipper this weekend.

Does the bananapi m2 zero fit into gpio pins on the einsy board?

Any chance you are up on discord and we could discuss further. I am a bit
of a novice and would appreciate a tiny bit of hand holding

**Reply by JesterIsDead** [reply]
The M2 is the same footprint with the same GPI0 lay out. I am on Discord. Jester #4835

**Comment by confuzled-dev** [comment]
Will do. However, it seems we need to be in a common discord server. Where
can I find you

**Reply by JesterIsDead** [reply]
You can just add me and call as well. I'm in the Klipper discord as well.
